AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginIntegrationTests/DataStoreObserveQueryTests.swift

```diff
@@ -104,11 +179,223 @@ class DataStoreObserveQueryTests: SyncEngineIntegrationTestBase {
+    /// Sync is completed and actions are applied after sync to observe expected snapshots.
+    ///
+    /// - Given: DataStore has already synced the models to local store. ObserveQuery is started with a predicate
+    /// - When/Then:
+    ///     - Add at least one model that matches the given predicate. Model is received on the snapshot.
+    ///     - Add at least one model that does NOT match the given predicate. Model is not received on the snapshot
+    ///     - Update a model that matches the given predicate so that it no longer matches the predicate.
+    ///         Model is removed from the snapshot.
+    ///     - Update a model that does NOT match the predicate so that it now matches the predicate.
+    ///         Model is added to the snapshot
+    ///     - Delete a model that matches the predicate. Model is removed from the snapshot
+    ///     - Delete a model that does NOT match the predicate. No snapshot is emitted
+    func testPredicateWithCreateUpdateDelete() throws {
+        let started = expectation(description: "Amplify started")
+        try startAmplify {
+            started.fulfill()
+        }
+        wait(for: [started], timeout: 2)
+        Amplify.Logging.logLevel = .info
+
+        let testId = UUID().uuidString
+        let postMatchPredicate = Post(title: "xyz 1", content: testId, createdAt: .now())
+        let postDoesNotMatch = Post(title: "doesNotMatch", content: testId, createdAt: .now())
+
+        var snapshotCount = 0
+        let allSnapshotsReceived = expectation(description: "query snapshots received")
+        let predicate = Post.keys.title.beginsWith("xyz") && Post.keys.content == testId
+        let sink = Amplify.DataStore.observeQuery(for: Post.self, where: predicate).sink { completed in
+            switch completed {
+            case .finished:
+                break
+            case .failure(let error):
+                XCTFail("\(error)")
+            }
+        } receiveValue: { querySnapshot in
+            snapshotCount += 1
+            let items = querySnapshot.items
+            switch snapshotCount {
+            case 1:
+                self.log.info("1. \(querySnapshot)")
+                XCTAssertEqual(items.count, 0)
+            case 2, 3:
+                // See (1), subsequent snapshot should have item with "xyz 1".
+                self.log.info("2/3. \(querySnapshot)")
+                XCTAssertEqual(items.count, 1)
+                XCTAssertEqual(items[0].title, "xyz 1")
+            case 4:
+                // See (2), should not be added to the snapshot.
+                // See (3), should be removed from the snapshot. So the resulting snapshot is empty.
+                self.log.info("4. \(querySnapshot)")
+                XCTAssertEqual(items.count, 0)
+            case 5, 6:
+                // See (4). the post that now matches the snapshot should be added
+                self.log.info("5/6. \(querySnapshot)")
```

Comment:
for update and predicate match, snapshots are received twice `case 5,6`
for update and predicate not matching, snapshot is received once `case 4` 
is this the case?

Reply:
> for update and predicate match, snapshots are received twice case 5,6

Yup because the update is added twice due to the locally source mutation and remote event

> for update and predicate not matching, snapshot is received once case 4

Yes because the item has been removed from the snapshot (due to locally source mutation), and the remote event will no longer be applied to the snapshot (since it does not exist in the snapshot anymore) so the second event does not cause a snapshot to be emitted a second time